### PR TITLE
refactor(lsp): refactor completions and add tests

### PR DESCRIPTION
**Pull request by kitsonk**
This refactor is to give us a more solid base to build on for import completions.

It refactors some code and adds a completion resolver, plus some tests.

### CODE REVIEW

File: cli/lsp/config.rs [hunk 1]
```diff
@@ -30,13 +30,50 @@ pub struct CodeLensSettings {
...
+}
+
 #[derive(Debug, Default, Clone, Deserialize)]
 #[serde(rename_all = "camelCase")]
 pub struct WorkspaceSettings {
   pub enable: bool,
   pub config: Option<String>,
   pub import_map: Option<String>,
-  pub code_lens: Option<CodeLensSettings>,
+  #[serde(default)]
```
Comment (kitsonk) [inline cli/lsp/config.rs:73]: These give us default values if they aren't sent by the client, this makes it a lot easier dealing with the values.

**Review by lucacasonato — APPROVED** [review]
Looks great. Tested it locally, and optional chaining completions now work correctly. :+1:

File: cli/lsp/tsc.rs [hunk 2]
```diff
@@ -348,42 +358,58 @@ pub enum ScriptElementKind {
...
-    use lspower::lsp::CompletionItemKind;
-
     match kind {
       ScriptElementKind::PrimitiveType | ScriptElementKind::Keyword => {
-        CompletionItemKind::Keyword
+        lsp::CompletionItemKind::Keyword
       }
-      ScriptElementKind::ConstElement => CompletionItemKind::Constant,
-      ScriptElementKind::LetElement
+      ScriptElementKind::ConstElement
```
[the remark this answers is not in the document]
Comment (kitsonk) [reply cli/lsp/tsc.rs:373]: Yes, I will double check that I didn't make a mistake though.

[the remark this answers is not in the document]
Comment (kitsonk) [reply cli/lsp/tsc.rs:373]: Nope, that is the way it is in VSCode 🤷 :

```ts
private static convertKind(kind: string): vscode.CompletionItemKind {
  switch (kind) {
    case PConst.Kind.primitiveType:
    case PConst.Kind.keyword:
      return vscode.CompletionItemKind.Keyword;

    case PConst.Kind.const:
    case PConst.Kind.let:
    case PConst.Kind.variable:
    case PConst.Kind.localVariable:
    case PConst.Kind.alias:
    case PConst.Kind.parameter:
      return vscode.CompletionItemKind.Variable;

    case PConst.Kind.memberVariable:
    case PConst.Kind.memberGetAccessor:
    case PConst.Kind.memberSetAccessor:
      return vscode.CompletionItemKind.Field;

    case PConst.Kind.function:
    case PConst.Kind.localFunction:
      return vscode.CompletionItemKind.Function;

    case PConst.Kind.method:
    case PConst.Kind.constructSignature:
    case PConst.Kind.callSignature:
    case PConst.Kind.indexSignature:
      return vscode.CompletionItemKind.Method;

    case PConst.Kind.enum:
      return vscode.CompletionItemKind.Enum;

    case PConst.Kind.enumMember:
      return vscode.CompletionItemKind.EnumMember;

    case PConst.Kind.module:
    case PConst.Kind.externalModuleName:
      return vscode.CompletionItemKind.Module;

    case PConst.Kind.class:
    case PConst.Kind.type:
      return vscode.CompletionItemKind.Class;

    case PConst.Kind.interface:
      return vscode.CompletionItemKind.Interface;

    case PConst.Kind.warning:
      return vscode.CompletionItemKind.Text;

    case PConst.Kind.script:
      return vscode.CompletionItemKind.File;

    case PConst.Kind.directory:
      return vscode.CompletionItemKind.Folder;

    case PConst.Kind.string:
      return vscode.CompletionItemKind.Constant;

    default:
      return vscode.CompletionItemKind.Property;
  }
}
```

**Pull request merged**